Sources/NIOCore/NIOIPAddresses.swift

```diff
@@ -0,0 +1,583 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+
+import Darwin
+import CNIOLinux
+import CoreFoundation
+
+
+/// Represent a single `NIOIPAddress`.
+public enum NIOIPAddress: CustomStringConvertible {
+    /// `Error` that may be thrown if we fail to create a `NIOIPAddress`
+    public enum IPAddressError: Error {
+        /// Given string input is not supported IP Address
+        case failedToParseIPString(String)
+        /// Given string input is not supported IP Address
+        case bytesArrayHasWrongLength(Int)
+    }
+
+    public typealias IPv4BytesTuple = (UInt8, UInt8, UInt8, UInt8)
+    public typealias IPv6BytesTuple = (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
+
+    /// Represent the bytes for an `IPv4Address`.
+    public struct IPv4Bytes {
+        public var bytes: IPv4BytesTuple
+        
+        public init(_ bytes: IPv4BytesTuple) {
+            self.bytes = bytes
+        }
+    }
+
+    /// Represent the bytes for an `IPv6Address.
+    public struct IPv6Bytes {
+        public var bytes: IPv6BytesTuple
+        
+        public init(_ bytes: IPv6BytesTuple) {
+            self.bytes = bytes
+        }
+    }
```

Comment:
These types have a collection conformance hung off them, while the IP addresses themselves do not. I think that's a reasonable API design choice.

Reply:
Is it the _right_ API design choice, though? Especially when you consider not just the type, but the extra conformances as well.

Reply:
Broadly my instinct is yes. Tuples are profoundly trash: their inability to expose a collection interface makes working with them a chore, their inability to be named or evolved makes them a trap for API evolution, and using them prominently in the API usually invites creating an alternative representation at some point down the road.

In the case of the extra type I consider that a plus, not a minus: nominal types are good for all the reasons tuples are bad. As for the conformances, I also don't consider conformances to be inherently a bad thing. Yes, they have a binary size impact, but for server applications that impact is substantially less important than it is for client use-cases, and in the case of this specific type the conformances are justified by the utility they bring (for example, allowing users to compute on the bytes of the tuple easily).

Reply:
If Swift had fixed-size arrays (i.e. tuples that conform to collection), that would be much better. In the case of IP addresses, I don't think evolution is really a problem though - and it is actually the role of the IP address type to be that wrapper over the tuple which gives its pieces meaning. If there is a compelling reason for a collection-of-bytes API, IMO it should be exposed as a 'view' of the IP address.

Another issue to be aware of is that storing these as tuples of bytes may cause alignment problems and other code-gen difficulties. I'm thinking about changing WebURL's implementation to store a `UInt32` and `(UInt64, UInt64)` to guarantee that operations such as equality and comparison lower to simple integer operations rather than expensive byte-wise comparison (it is surprisingly difficult to do that).

Switching between stored and computed properties is API compatible so this is just something to be aware of, although the required changes would be significant so may be worth doing it now.

Reply:
> In the case of IP addresses, I don't think evolution is really a problem though

API evolution is basically always a problem. It's very hard to know when a language feature is going to come along that you'd like to adopt.

> and it is actually the role of the IP address type to be that wrapper over the tuple which gives its pieces meaning. If there is a compelling reason for a collection-of-bytes API, IMO it should be exposed as a 'view' of the IP address.

I agree, and IMO that's the shape of the current API. The name of the accessor to the view is `.address`, which is bad, but if it were `.bytes` then I think the API would be exactly what you're looking for.

> WebURL exposes the bytes using the tuple elements, and it works well IMO. I'm not sure full Collection conformance is totally necessary beyond this, but maybe you think it is?

I do: a bunch of APIs are spelled in terms of Collection and related protocols, and it's nice to be able to pass the type directly in. Tuples either require wrapping or unsafe code: it's nicer for us to write the few lines of code required to do the wrapping once. In this instance we aren't forbidding access to the tuple: we're just prioritising access as a `Collection`.

As to the layout: yes, I agree, we could swap to using other representations (and my original discussion of the issue proposed that representation), but we can straightforwardly change it later so I'm not too fussed right now.

Reply:
`.bytes` view sounds good!